janus_server/src/datastore.rs

```diff
@@ -1239,6 +1239,118 @@ impl<C: Clock> Transaction<'_, C> {
         Ok(collect_job_id)
     }
 
+    /// acquire_incomplete_collect_jobs retrieves & acquires the IDs of unclaimed incomplete collect
+    /// jobs. At most `maximum_acquire_count` jobs are acquired. The job is acquired with a "lease"
+    /// that will time out; the desired duration of the lease is a parameter, and the lease
+    /// expiration time is returned.
+    pub async fn acquire_incomplete_collect_jobs(
+        &self,
+        lease_duration: Duration,
+        maximum_acquire_count: usize,
+    ) -> Result<Vec<(AcquiredCollectJob, Time)>, Error> {
+        let now = self.clock.now();
+        let lease_expiry_time = now.add(lease_duration)?;
+        let maximum_acquire_count: i64 = maximum_acquire_count.try_into()?;
+
+        let stmt = self
+            .tx
+            .prepare_cached(
+               "WITH updated as (
+                    UPDATE collect_jobs SET lease_expiry = $1
+                    FROM collect_jobs as collect_jobs_alias
+                    -- Join on tasks table
+                    INNER JOIN tasks ON tasks.id = collect_jobs_alias.task_id
+                    -- Join on aggregation jobs with matching task ID and aggregation parameter
+                    INNER JOIN aggregation_jobs
+                        ON collect_jobs_alias.aggregation_param = aggregation_jobs.aggregation_param
+                        AND collect_jobs_alias.task_id = aggregation_jobs.task_id
+                    -- Join on report aggregations with matching aggregation job ID
+                    INNER JOIN report_aggregations
+                        ON report_aggregations.aggregation_job_id = aggregation_jobs.id
+                    -- Join on reports whose nonce falls within the collect job batch interval and which
+                    -- are included in an aggregation job
+                    INNER JOIN client_reports
+                        ON client_reports.nonce_time <@ tsrange(
+                            collect_jobs_alias.batch_interval_start,
+                            collect_jobs_alias.batch_interval_start + collect_jobs_alias.batch_interval_duration * interval '1 second')
+                        AND client_reports.id = report_aggregations.client_report_id
+                    WHERE
+                        -- Do not acquire collect jobs that have been completed
+                        collect_jobs_alias.helper_aggregate_share IS NULL
+                        -- Do not acquire collect jobs with an unexpired lease
+                        AND collect_jobs_alias.lease_expiry <= $2
+                        -- Filter out collect jobs where any associated aggregation jobs are not
+                        -- finished
+                        -- XXX How to achieve this?
+                        -- AND collect_jobs_alias.id NOT IN (SELECT collect_jobs_alias.id FROM updated WHERE aggregation_jobs.state = 'IN_PROGRESS')
```

Comment:
I need some help with this. This big series of joins figures out which aggregation jobs are related to report shares whose nonce is within the collect job's batch interval. There may be multiple aggregation jobs for a single collect job. I can't quite figure out how to handle the case where some but not all of the aggregation jobs are in the `FINISHED` state. Basically, if there are any rows in the giant joined table where `aggregation_jobs.state` is `IN_PROGRESS`, then I want to remove all other rows that have the same `collect_jobs_alias.id`. It feels like I want to intersect the table with itself, and I'm not sure how to do that or if it's a good idea.

Reply:
I need to chew on this some more, but my idea is to put a SELECT subquery in the UPDATE's FROM position, use a GROUP BY in the SELECT, and impose our for-all conditions with the bool_and aggregate function. I think we definitely need to make use of subqueries somehow here, because our UPDATE query may otherwise return multiple rows per collect_job row. Per the [Postgres docs](https://www.postgresql.org/docs/14/sql-update.html#id-1.9.3.182.8): "When using FROM you should ensure that the join produces at most one output row for each row to be modified."

Other notes: I think we don't want `FROM collect_jobs as collect_jobs_alias`, as that gives us a self-join on that table, once from the UPDATE table, and once from the FROM clause. The `LIMIT $3` needs to be in the UPDATE subquery, rather than the wrapping `SELECT`, to keep us from taking out extra leases and dropping them on the ground.

Reply:
Okay, I got the test passing at least: 5a2c0d07714c48ee8d46fdbb45c42631296b57ad

Reply:
THANK YOU! I was going nuts trying to figure this out. Not to mention the other two or three crippling bugs you spotted!

I did have to make one tweak, to move the `WHERE` clauses into the inner `SELECT` where the `LIMIT` is applied. Otherwise, the `LIMIT` gets applied to query results before we check for collect job lease expiry or completion, and so we might end up discarding runnable jobs.